### PR TITLE
618 withjson support method for put index template request

### PR DESCRIPTION
**Pull request by pranishd1**
### Description
Added "withJson" support for PutIndexTemplateRequest

### Issues Resolved
Closes [#618]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — COMMENTED** [review]
There are a lot of requests that support JSON bodies. I think we need to introduce a class or train that supports `withJson` at a lower level than `PutIndexTemplateRequest` and then include that in it. WDYT?

**Comment by pranishd1** [comment]
Yes, that would be great. I will look into that.

**Comment by VachaShah** [comment]
@pranishd1 Can you fix merge conflicts please? Also, you might have to run `./gradlew spotlessApply` to fix formatting.

File: java-client/src/main/java/org/opensearch/client/util/WithJsonBuilderBase.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
```
[the remark this answers is not in the document]
Comment (pranishd1) [reply java-client/src/main/java/org/opensearch/client/util/WithJsonBuilderBase.java]: Hi. I have removed the ES license header from the newer file. My understanding was that all files should have consistent headers.

**Review by VachaShah — COMMENTED** [review]
Can we also mention in the USER_GUIDE that how to use `withJson`?

**Comment by dblock** [comment]
This is a lot better! I'd love a guide (maybe copy from https://github.com/opensearch-project/opensearch-py/pull/542?) and a working sample as part of this PR, please?

**Comment by pranishd1** [comment]
> Can we also mention in the USER_GUIDE that how to use `withJson`?

sure, I will add a section about how to use this to the user guide.

**Comment by pranishd1** [comment]
> This is a lot better! Can we apply this to at least some minimum number of requests, not just put index template? I'd also love a guide (maybe copy from [opensearch-project/opensearch-py#542](https://github.com/opensearch-project/opensearch-py/pull/542)?) and a working sample as part of this PR, please?

Yeah, sure I will add this to a few other requests. Thanks for the feedback.

**Comment by pranishd1** [comment]
@dblock , sorry about the delay. I am in the process of implementing the interface in create, search and delete requests. While doing so, I am stuck at object deserialization process. [This method](https://github.com/opensearch-project/opensearch-java/pull/657/files#:~:text=default%20B-,withJson,-(JsonParser%20parser)) does not cover all the objects.

**Comment by Jai2305** [comment]
Thanks for this guys, much needed, do you need any help with this? 
Also after reading the code I got to know the problem - 
that as of now we rely on static deserializer which creates a new builder everytime, referring to this code exactly -
``` Java
public static final JsonpDeserializer<SearchRequest> _DESERIALIZER = ObjectBuilderDeserializer.lazy(
        Builder::new,
        SearchRequest::setupSearchRequestDeserializer
    );
```

Although some suggestions -
* Can you please also create a method which takes Readable as an argument kind of like 
``` Java
 default<T> B withJson(Reader reader) {
             ........ 
```
* Also now that we have default mapper in JsonpUtils, we can use that rather than relying on a specific implementation 
please refer [here ](https://github.com/opensearch-project/opensearch-java/pull/1064/files#diff-75ccc5d58d9779725e7becbdfa9c4a6dfeee3772bc28a93b92721a3ef1f6e9adR71:~:text=static%20final%20JsonpMapper%20DEFAULT_JSONP_MAPPER%20%3D%20new%20JsonpMapperBase()%20%7B).

**Comment by dblock** [comment]
@Jai2305 looks like the PR hasn't been touched in a long time, so yes, would love it if you could pick it up and finish it!

Note that we added support for a generic interface since which takes/returns raw JSON, so some parts of that may be reusable.

**Comment by Xtansia** [comment]
A version of this was added in #1148, documented here: https://github.com/opensearch-project/opensearch-java/blob/main/guides/json.md#using-withjson

**Pull request closed without merging**